tests/trace_server/test_opentelemetry.py

```diff
@@ -0,0 +1,426 @@
+from weave.trace_server.opentelemetry.attributes import (
+    expand_attributes,
+    flatten_attributes,
+    get_attribute,
+    to_json_serializable,
+    unflatten_key_values,
+)
+from weave.trace_server.opentelemetry.python_spans import (
+    SpanKind,
+    StatusCode,
+)
+from weave.trace_server.opentelemetry.python_spans import Span as PySpan
+from weave.trace_server.opentelemetry.python_spans import TracesData as PyTracesData
+
+
+def create_test_span():
+    """Create a test OpenTelemetry Span."""
+    span = Span()
+    span.name = "test_span"
+    span.trace_id = uuid.uuid4().bytes
+    span.span_id = uuid.uuid4().bytes[:8]
+    span.start_time_unix_nano = int(datetime.now().timestamp() * 1_000_000_000)
+    span.end_time_unix_nano = span.start_time_unix_nano + 1_000_000_000  # 1 second later
+    span.kind = 1
+
+    # Add some attributes
+    kv1 = KeyValue()
+    kv1.key = "test.attribute"
+    kv1.value.string_value = "test_value"
+    span.attributes.append(kv1)
+
+    kv2 = KeyValue()
+    kv2.key = "test.number"
+    kv2.value.int_value = 42
+    span.attributes.append(kv2)
+
+    # Create nested attributes
+    kv3 = KeyValue()
+    kv3.key = "test.nested.value"
+    kv3.value.string_value = "nested_test_value"
+    span.attributes.append(kv3)
+
+    # Create an array attribute
+    kv4 = KeyValue()
+    kv4.key = "test.array"
+    array_value = AnyValue()
+    value1 = AnyValue()
+    value1.string_value = "value1"
+    value2 = AnyValue()
+    value2.string_value = "value2"
+    array_value.array_value.values.extend([value1, value2])
+    kv4.value.CopyFrom(array_value)
+    span.attributes.append(kv4)
+
+    # Set status
+    span.status.code = StatusCode.OK.value
+    span.status.message = "Success"
+
+    return span
+
+
+def create_test_export_request(project_id="test_project"):
+    """Create a test ExportTraceServiceRequest with one span."""
+    span = create_test_span()
+
+    # Create instrumentation scope
+    scope = InstrumentationScope()
+    scope.name = "test_instrumentation"
+    scope.version = "1.0.0"
+
+    # Create scope spans
+    scope_spans = ScopeSpans()
+    scope_spans.scope.CopyFrom(scope)
+    scope_spans.spans.append(span)
+
+    # Create resource with attributes
+    resource = Resource()
+    kv = KeyValue()
+    kv.key = "service.name"
+    kv.value.string_value = "test_service"
+    resource.attributes.append(kv)
+
+    # Create resource spans
+    resource_spans = ResourceSpans()
+    resource_spans.resource.CopyFrom(resource)
+    resource_spans.scope_spans.append(scope_spans)
+
+    # Create traces data
+    traces_data = TracesData()
+    traces_data.resource_spans.append(resource_spans)
+
+    # Create export request
+    request = ExportTraceServiceRequest()
+    request.resource_spans.append(resource_spans)
+
+    return tsi.OtelExportReq(
+        project_id=project_id,
+        traces=request,
+        wb_user_id=None
+    )
+
+
+@pytest.fixture
+def mock_clickhouse_trace_server():
```

Comment:
I think you instead want to use the `client` fixture. These mocks seem to hide some of the critical things you are testing.

Reply:
Should I use test_custom_provider as inspiration? Or is there another example you recommend

Reply:
there are very many such examples see `test_trace_client`. when you use the `client` fixture, you get `client.server` which is basically your clickhouse batched server